packages/gatsby-theme-docs/src/modules/sso/hooks/use-authentication.ts

```diff
@@ -0,0 +1,70 @@
+import { useContext, useEffect, useState } from 'react';
+import useSWR from 'swr';
+import { useAuth0 } from '@auth0/auth0-react';
+import { useAuthToken } from '../../self-learning/hooks/use-auth-token';
+import ConfigContext from '../../../components/config-context';
+import { fetcherWithToken } from '../../self-learning/hooks/hooks.utils';
+import { getCookieValue } from '../utils/common.utils';
+
+const doesSessionExist = (cookieContent?: string): boolean => {
+  if (!cookieContent) {
+    return false;
+  }
+  // Parse the JSON content of the cookie
+  try {
+    const parsedContent = JSON.parse(decodeURIComponent(cookieContent));
+    return parsedContent?.sessionExists;
+  } catch (err) {
+    return false;
+  }
+};
+
+const useAuthentication = () => {
+  const { isAuthenticated: isAuth0Authenticated, ...rest } = useAuth0();
+  const [isAuthenticated, setIsAuthenticated] = useState<boolean>(false);
+  const [shouldInitSession, setShouldInitSession] = useState(false);
+
+  const { learnApiBaseUrl } = useContext(ConfigContext);
+  const { getAuthToken } = useAuthToken();
+
+  const apiEndpoint = `/api/init-session`;
+
+  const { data, error, isLoading } = useSWR(
+    shouldInitSession ? apiEndpoint : null,
+    (url) => fetcherWithToken(url, getAuthToken, learnApiBaseUrl, 'POST'),
+    {
+      revalidateOnReconnect: false,
+      revalidateIfStale: false,
+      revalidateOnMount: true,
+      revalidateOnFocus: false,
+    }
+  );
+
+  useEffect(() => {
+    const checkUserSessionCookie = () => {
+      const cookieName = 'user_session_for_client';
+      const cookieValue = getCookieValue(cookieName);
+      if (!doesSessionExist(cookieValue)) {
+        setShouldInitSession(true);
+      } else {
+        setIsAuthenticated(true);
```

Comment:
it may be a nitpick or I am not fully understanding the logic here.  The fact that a session from the learn API exists does not imply that the user is also authenticated auth0-side.   We try to clear the session when logging out but Auth0 can also log out or expire a session in other ways.  

or is this just the wrapper logic that "delays" the signal to the rest of the app that the user is authenticated until the moodle API session is there, too?

Reply:
Hi @nkuehn , 
I think you might be mislead by the function closure. `checkUserSessionCookie` gets invoked only if `isAuth0Authenticated === true` (see below). The `useEffect` is triggered only on `isAuth0Authenticated` change AND, only when it's true, the check session happens.